src/azul/terra.py

```
@@ -511,11 +513,26 @@ def run_sql(self, query: str) -> BigQueryRows:
                                     attempt + 1, len(delays) + 1, delay, exc_info=e)
                         sleep(delay)
                 else:
-                    break
+                    if delay is None:
+                        raise ServiceUnavailable('Exceeded number of attempts')
```

This check happens too early. This way, we will _always_ raise an exception on the 5th retry, regardless of whether the job silently failed or not, potentially throwing away a valid result. I also don't think we should use an exception class from the BigQuery package here because it could be mistaken for an exception raised from within the package code.

---

At minimum, I would do:

```diff
Subject: [PATCH] REVIEW
---
Index: src/azul/terra.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/terra.py b/src/azul/terra.py
--- a/src/azul/terra.py	(revision 2dfb45efc3453530acba88c9c736b14af26af62c)
+++ b/src/azul/terra.py	(date 1744916478050)
@@ -513,8 +513,6 @@
                                     attempt + 1, len(delays) + 1, delay, exc_info=e)
                         sleep(delay)
                 else:
-                    if delay is None:
-                        raise ServiceUnavailable('Exceeded number of attempts')
                     # Don't exit the retry loop if the BigQuery job silently
                     # failed. Note, this check can't be performed in a test
                     # environment since the job stats can be empty when using
@@ -524,9 +522,12 @@
                         'MockTDRClient' not in str(self)
                         and 'totalBytesProcessed' not in json_mapping(job_info['stats'])
                     ):
-                        log.warning('BigQuery job silent fail during attempt %i/%i. Retrying in %is.',
-                                    attempt + 1, len(delays) + 1, delay)
-                        sleep(delay)
+                        if delay is None:
+                            raise ServiceUnavailable('Exceeded number of attempts')
+                        else:
+                            log.warning('BigQuery job silent fail during attempt %i/%i. Retrying in %is.',
+                                        attempt + 1, len(delays) + 1, delay)
+                            sleep(delay)
                     else:
                         break
             else:
```

---

This is more experimental and may be contentious, but eliminates the repetitive log messages and delay checks. See what you think. It requires creating a custom exception class.

```diff
Subject: [PATCH] REVIEW
---
Index: src/azul/terra.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/terra.py b/src/azul/terra.py
--- a/src/azul/terra.py	(revision 2dfb45efc3453530acba88c9c736b14af26af62c)
+++ b/src/azul/terra.py	(date 1744921588580)
@@ -500,7 +500,10 @@
                 job = bigquery.query(query)
                 try:
                     result = job.result()
-                except (BadRequest, Forbidden, InternalServerError, ServiceUnavailable) as e:
+                    job_info = self._job_info(job, result)
+                    if not self._job_has_result(job_info):
+                        raise EmptySQLResult
+                except (BadRequest, Forbidden, InternalServerError, ServiceUnavailable, EmptySQLResult) as e:
                     if delay is None:
                         raise e
                     elif isinstance(e, Forbidden) and 'Exceeded rate limits' not in e.message:
@@ -511,24 +514,9 @@
                     else:
                         log.warning('BigQuery job error during attempt %i/%i. Retrying in %is.',
                                     attempt + 1, len(delays) + 1, delay, exc_info=e)
-                        sleep(delay)
-                else:
-                    if delay is None:
-                        raise ServiceUnavailable('Exceeded number of attempts')
-                    # Don't exit the retry loop if the BigQuery job silently
-                    # failed. Note, this check can't be performed in a test
-                    # environment since the job stats can be empty when using
-                    # the MockTDRClient.
-                    job_info = self._job_info(job, result)
-                    if (
-                        'MockTDRClient' not in str(self)
-                        and 'totalBytesProcessed' not in json_mapping(job_info['stats'])
-                    ):
-                        log.warning('BigQuery job silent fail during attempt %i/%i. Retrying in %is.',
-                                    attempt + 1, len(delays) + 1, delay)
                         sleep(delay)
-                    else:
-                        break
+                else:
+                    break
             else:
                 assert False
         if log.isEnabledFor(logging.DEBUG):
```